### PR TITLE
JDK-8285730: unify _WIN32_WINNT settings

### PR DESCRIPTION
**Pull request by MBaesken**
Currently we set _WIN32_WINNT at various places in the codebase; this is used to target a minimum Windows version we want to support. See also for more detailled information :
https://docs.microsoft.com/en-us/windows/win32/winprog/using-the-windows-headers?redirectedfrom=MSDN#setting-winver-or-_win32_winnt
Macros for Conditional Declarations
"Certain functions that depend on a particular version of Windows are declared using conditional code. This enables you to use the compiler to detect whether your application uses functions that are not supported on its target version(s) of Windows."

However currently we have quite a lot of differing settings of _WIN32_WINNT in the codebase ; setting _WIN32_WINNT to 0x0601 (Windows 7) where possible would make sense because we have this setting already at   java_props_md.c  (so targeting older Windows versions at other places is most likely not useful).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8285730](https://bugs.openjdk.java.net/browse/JDK-8285730): unify _WIN32_WINNT settings

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**)
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**)
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8428/head:pull/8428` \
`$ git checkout pull/8428`

Update a local copy of the PR: \
`$ git checkout pull/8428` \
`$ git pull https://git.openjdk.java.net/jdk pull/8428/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8428`

View PR using the GUI difftool: \
`$ git pr show -t 8428`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8428.diff">https://git.openjdk.java.net/jdk/pull/8428.diff</a>

</details>

### CODE REVIEW

File: src/java.base/windows/native/libnio/ch/wepoll.c [hunk 1]
```diff
@@ -156,7 +156,7 @@ WEPOLL_EXPORT int epoll_wait(HANDLE ephnd,
 #define WIN32_LEAN_AND_MEAN
 
 #undef _WIN32_WINNT
-#define _WIN32_WINNT 0x0600
+#define _WIN32_WINNT 0x0601
```
Comment (AlanBateman) [inline src/java.base/windows/native/libnio/ch/wepoll.c:159]: This is 3rd party code and would prefer not change it if possible.

**Comment by prrace** [comment]
This is OK by me, but I wonder if the build folks might want to think about this and whether it should be centralised somehow
since it seems odd to mandate different versions of windows in different places.

**Comment by erikj79** [comment]
Requiring different windows versions in different parts of the product doesn't make much sense, but I think it's even worse trying to keep all these different locations in sync. At least most of these have a comment explaining why that particular Windows version is required there. Changing these values invalidates those comments. 

If we are to do something about this, then we need to define this macro in a central location, and verify the value in configure. Then we would provide it either on compile command lines, or in a globally included config.h.

**Comment by dholmes-ora** [comment]
There is obviously a lot of history here and different parts of the codebase had hit different minimum OS version requirements at different times. While at one time we would have had to cater for building on systems with and without a given API those days are long gone for the code being touched here (AFAICS). We should be able to set a minimum _WIN32_WINNT_ value needed to build all the codebase and simply have that checked and set at configure time. The code itself would not need to define _WIN32_WINNT.

File: src/java.base/windows/native/libnio/ch/wepoll.c (hunk 1 above)
Comment (MBaesken) [reply src/java.base/windows/native/libnio/ch/wepoll.c:159]: I Alan, I agree (thats why I did not change the define in harfbuzz,  but I missed that wepoll.c is 3rd party code as well).

**Comment by MBaesken** [comment]
Hi Erik/David/Phil, we already have a good central place where we set the definition of WIN32_LEAN_AND_MEAN 

autoconf/flags-cflags.m4:460:    ALWAYS_DEFINES_JDK="-DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE \
autoconf/flags-cflags.m4:462:    ALWAYS_DEFINES_JVM="-DNOMINMAX -DWIN32_LEAN_AND_MEAN"

should we add there the _WIN32_WINNT=0x0601  define  (and remove the differing defines instead at the other places) ?
(defines of _WIN32_WINNT in 3rd party code would stay like wepoll.c and harfbuzz).

**Comment by erikj79** [comment]
> Hi Erik/David/Phil, we already have a good central place where we set the definition of WIN32_LEAN_AND_MEAN
> 
> autoconf/flags-cflags.m4:460: ALWAYS_DEFINES_JDK="-DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE autoconf/flags-cflags.m4:462: ALWAYS_DEFINES_JVM="-DNOMINMAX -DWIN32_LEAN_AND_MEAN"
> 
> should we add there the _WIN32_WINNT=0x0601 define (and remove the differing defines instead at the other places) ? (defines of _WIN32_WINNT in 3rd party code would stay like wepoll.c and harfbuzz).

Seems reasonable to me at least.

**Comment by dholmes-ora** [comment]
That only seems to be half of the issue though. If we are defining _WIN32_WINNT=0x0601 because the minimum required OS API support level is Windows 7, then don't we need a check that the build platform is also at least Windows 7?

**Comment by MBaesken** [comment]
> That only seems to be half of the issue though. If we are defining _WIN32_WINNT=0x0601 because the minimum required OS API support level is Windows 7, then don't we need a check that the build platform is also at least Windows 7?

Hi David,  on older OS than Windows 7 we wouldn't be able to build OJDK anyway currently. Our oldest VS we still support (see toolchain_microsoft.m4) is VS2017.  
VS2017 has the following system requirements 
https://docs.microsoft.com/en-us/visualstudio/releases/2017/vs2017-system-requirements-vs
see supported OS ,  there the oldest supported OS is Windows Server 2012 R2 and Windows 7 SP1. So on older than Win7 we would not even have a compiler anymore that passes configure.

**Comment by erikj79** [comment]
If we define this centrally using compiler flags, then we should not also update each location in the source. Those need to either be removed or left alone. In the cases where the source definition is guarded with an ifndef and there is a comment explaining why a particular version of windows is required, keeping it around could make sense. But on the other hand, since those defines are overridden using flags, they are likely to bit rot over time so we might as well just remove all of them.

**Comment by dholmes-ora** [comment]
I agree with Erik - the source locations need to be modified to not define it. If we want to keep a record perhaps add an assertion that the value is what was expected?

I still feel we have a disconnect between this and an actual check for what the build and runtime platform version is ...

and it isn't at all clear how someone using an API only in a later Windows version, and so setting _WIN32_WINNT to a higher value, will know that this is defined down in the build files ?

File: src/java.base/windows/native/libnio/ch/wepoll.c (hunk 1 above)
Comment (MBaesken) [reply src/java.base/windows/native/libnio/ch/wepoll.c]: Yes sure I did it .

**Comment by MBaesken** [comment]
Interesting fact :
we run now into this compile error :

d:\a\jdk\jdk\jdk\src\jdk.crypto.mscapi\windows\native\libsunmscapi\security.cpp(1262): error C2065: 'NCRYPT_CIPHER_KEY_BLOB': undeclared identifier
d:\a\jdk\jdk\jdk\src\jdk.crypto.mscapi\windows\native\libsunmscapi\security.cpp(1280): error C2065: 'NCRYPT_PROTECTED_KEY_BLOB': undeclared identifier

Reason seems that already for some time  ( at least OpenJDK 11 (!) )  we have an implicit minimum requirement of Windows 8 / Windows 2012
APIs for this code, so enforcing Win 7 is too old :

https://docs.microsoft.com/en-us/windows/win32/api/ncrypt/nf-ncrypt-ncryptexportkey

NCRYPT_PROTECTED_KEY_BLOB
Export a protected key in a NCRYPT_KEY_BLOB_HEADER structure.

Windows 8 and Windows Server 2012:  Support for this value begins.

NCRYPT_CIPHER_KEY_BLOB
Export a cipher key in a NCRYPT_KEY_BLOB_HEADER structure.

Windows 8 and Windows Server 2012:  Support for this value begins.

**Comment by MBaesken** [comment]
> I agree with Erik - the source locations need to be modified to not define it. If we want to keep a record perhaps add an assertion that the value is what was expected?
> 
> I still feel we have a disconnect between this and an actual check for what the build and runtime platform version is ...
> 
> and it isn't at all clear how someone using an API only in a later Windows version, and so setting _WIN32_WINNT to a higher value, will know that this is defined down in the build files ?

Hi David, I agree the other locations as long as they are not 3rd party code should be cleaned up.
Someone using an API  that is only available in later Windows versions would get  a redefinition warning as long as to undefine is done.
But we have to set _WIN32_WINNT anyway to a higher value (windows8, I think that's 0x0602)  to  compile 
 src\jdk.crypto.mscapi\windows\native\libsunmscapi\security.cpp .

**Comment by dholmes-ora** [comment]
I'm confused.  `src\jdk.crypto.mscapi\windows\native\libsunmscapi\security.cpp` doesn't set _WIN32_WINNT so how is that later API being enabled? Does this mean that not setting _WIN32_WINNT means :any API is allowed" ?

**Comment by MBaesken** [comment]
> I'm confused. `src\jdk.crypto.mscapi\windows\native\libsunmscapi\security.cpp` doesn't set _WIN32_WINNT so how is that later API being enabled? Does this mean that not setting _WIN32_WINNT means :any API is allowed" ?

I found this info  here  https://stackoverflow.com/questions/37668692/visual-studio-setting-winver-win32-winnt-to-windows-8-on-windows-7
"VS 2012 uses the Windows 8.0 SDK which defaults to _WIN32_WINNT=0x0602 (Windows 8). VS 2013 / 2015 uses the Windows 8.1 SDK which defaults to _WIN32_WINNT=0x0603 (Windows 8.1). If you use VS 2015 and the Windows 10 SDK, it defaults to _WIN32_WINNT=0x0A00 (Windows 10)."

So it seems you get some more or less recent default when working with a not too old compiler/SDK.

**Comment by MBaesken** [comment]
> Does this mean that not setting _WIN32_WINNT means :any API is allowed" ?

Hi David , I did one more try with my current setup (VS2017 on a Win10 notebook). I did not set _WIN32_WINNT.
My little test program 

```
#include <windows.h>
#include <stdio.h>

int main() {

#ifdef _WIN32_WINNT
  printf("_WIN32_WINNT is defined .\n");

#if (_WIN32_WINNT == 0x0600)
  printf("Vista API setting\n");
#endif

#if (_WIN32_WINNT == 0x0601)
  printf("Win 7 API setting\n");
#endif

#if (_WIN32_WINNT == 0x0602)
  printf("Win 8 API setting\n");
#endif

#if (_WIN32_WINNT == 0x0603)
  printf("Win 8.1 API setting\n");
#endif

#if (_WIN32_WINNT == 0x0A00)
  printf("Win 10 API setting\n");
#endif

#endif

  return 0;
}
```

shows me 
_WIN32_WINNT is defined .
Win 10 API setting

So I think with our current compilers is use like VS2017 / VS2019 we allow Win10 APIs   in most of our code except a few places where we set _WIN32_WINNT  and go back to some mixture of older APIs.
Not sure if this is a good thing,  we could break for example Win 8.1/Win2012 compatibility easily this way.

**Comment by MBaesken** [comment]
Hi David, thanks for the review !

**Review by magicus — APPROVED** [review]
Looks good, thanks for doing this cleanup.

**Review by AlanBateman — APPROVED** [review]

**Comment by mkartashev** [comment]
This change seem to have made this group of declarations obsolete: 
`src/java.desktop/windows/native/libawt/windows/awt_Toolkit.h:157` (follow the [link](
https://github.com/openjdk/jdk/blob/89de756ffbefac452c7df559e2a4eb50bf71368b/src/java.desktop/windows/native/libawt/windows/awt_Toolkit.h#L157)). Does anyone have plans to drop those? Have any bugs been filed for that? If not, I'll attend to that myself.

**Comment by dholmes-ora** [comment]
@mkartashev  all references to WINVER in the AWT code seem obsolete. Possibly most of the IS_WINxxx usages could also be replaced.

**Comment by MBaesken** [comment]
> This change seem to have made this group of declarations obsolete: `src/java.desktop/windows/native/libawt/windows/awt_Toolkit.h:157` (follow the [link](https://github.com/openjdk/jdk/blob/89de756ffbefac452c7df559e2a4eb50bf71368b/src/java.desktop/windows/native/libawt/windows/awt_Toolkit.h#L157)). Does anyone have plans to drop those? Have any bugs been filed for that? If not, I'll attend to that myself.

Hi Maxim, no plans from me, so feel free to do further cleanups.

**Comment by mkartashev** [comment]
@dholmes-ora @MBaesken Thank you! Filed [JDK-8286634](https://bugs.openjdk.java.net/browse/JDK-8286634).

**Pull request closed without merging**